assets/src/components/carousel/index.js

```diff
@@ -0,0 +1,281 @@
+ */
+import { useRef, useEffect, useState, useCallback, useLayoutEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useWindowWidth } from '../../utils/use-window-width';
+import { CarouselNav } from './carousel-nav';
+
+export const DEFAULT_GUTTER_WIDTH = 60;
+export const DEFAULT_ITEM_WIDTH = 268;
+export const DEFAULT_MOBILE_BREAKPOINT = 783;
+
+/**
+ * Renders a scrollable carousel with a button navigation.
+ *
+ * @param {Object} props Component props.
+ * @param {number} props.gutterWidth Amount of space between items in pixels.
+ * @param {Array}  props.items Items in the carousel.
+ * @param {number} props.mobileBreakpoint Breakpoint below which to render the mobile version.
+ * @param {string} props.namespace CSS namespace.
+ * @param {number} props.highlightedItemIndex Index of an item receiving special visual treatment.
+ */
+export function Carousel( {
+	gutterWidth = DEFAULT_GUTTER_WIDTH,
+	items,
+	mobileBreakpoint = DEFAULT_MOBILE_BREAKPOINT,
+	namespace = 'amp-carousel',
+	highlightedItemIndex = 0,
+} ) {
+	const windowWidth = useWindowWidth();
+	const [ currentPage, originalSetCenteredPage ] = useState( null );
+	const [ pageWidth, setPageWidth ] = useState( 0 );
+	const carouselContainerRef = useRef();
+	const carouselListRef = useRef();
+
+	/**
+	 * Sets the the currentPage state and optionally scrolls to it.
+	 *
+	 * This state-setting wrapper is required, as opposed to scrolling to items in an effect hook when they're set as current,
+	 * because scroll listener below needs to set currentPage, but it does so only when the new currentPage is already centered
+	 * in the view. Calling scrollTo in that situation would cause jerky scroll effects.
+	 */
+	const setCurrentPage = useCallback( ( newCurrentPage, scrollToItem = true, smooth = true ) => {
+		originalSetCenteredPage( newCurrentPage );
+
+		if ( newCurrentPage && scrollToItem ) {
+			carouselListRef.current.scrollTo( { top: 0, left: newCurrentPage.offsetLeft, behavior: smooth ? 'smooth' : 'auto' } );
+		}
+	}, [] );
+
+	/**
+	 * Scroll to the highlighted item when it changes.
+	 */
+	useEffect( () => {
+		const newCurrentPage = carouselListRef.current.children.item( highlightedItemIndex );
+
+		setCurrentPage( newCurrentPage, true, false );
+	}, [ highlightedItemIndex, pageWidth, setCurrentPage ] );
+
+	/**
+	 * Set up a scroll listener to set an item as the currentPage as it crosses the center of the view.
+	 */
+	useLayoutEffect( () => {
+		let mounted = true;
+		const currentCarouselList = carouselListRef.current;
+
+		const scrollCallback = debounce( () => {
+			if ( ! mounted ) {
+				return;
+			}
+			for ( const child of [ ...currentCarouselList.children ] ) {
+				if ( child.offsetLeft >= currentCarouselList.scrollLeft ) {
+					if ( child !== currentPage ) {
+						setCurrentPage( child, false );
+					}
+					return;
+				}
+			}
+		}, 300 );
+		currentCarouselList.addEventListener( 'scroll', scrollCallback );
```

Comment:
```suggestion
		currentCarouselList.addEventListener( 'scroll', scrollCallback, { passive: true } );
```

See https://iamakulov.com/notes/resize-scroll/